src/fidesops/service/saas_request/saas_request_override_factory.py

```diff
@@ -0,0 +1,159 @@
+import logging
+from enum import Enum
+from inspect import Signature, signature
+from typing import Callable, Dict, List, Union
+
+from fidesops.common_exceptions import (
+    InvalidSaaSRequestOverrideException,
+    NoSuchSaaSRequestOverrideException,
+)
+from fidesops.util.collection_util import Row
+
+logger = logging.getLogger(__name__)
+
+# at some point this should likely be formalized more centrally...
+class SaaSRequestType(Enum):
+    """
+    An `Enum` containing the different possible types of SaaS requests
+    """
+
+    READ = "read'"
+    UPDATE = "update"
+    DATA_PROTECTION_REQUEST = "data_protection_request"
+    DELETE = "delete"
+
+
+class SaaSRequestOverrideFactory:
+    """
+    Factory class responsible for registering, maintaining, and providing
+    user-defined functions that act as overrides to SaaS request execution
+    """
+
+    registry: Dict[
+        SaaSRequestType, Dict[str, Callable[..., Union[List[Row], int]]]
+    ] = {}
+    valid_overrides: Dict[SaaSRequestType, str] = {}
+
+    # initialize each request type's inner dicts with an empty dict
+    for request_type in SaaSRequestType:
+        registry[request_type] = {}
+        valid_overrides[request_type] = ""
+
+    @classmethod
+    def register(
+        cls, name: str, request_types: List[SaaSRequestType]
```

Comment:
Is `request_types` a list for flexibility? I can't think of a use-case for this right now but it's better to have this just in case right?

Reply:
yup. the use case i could imagine was _potentially_ around update/delete? probably wouldn't make sense at this point, but just thought i'd keep things flexible.